### PR TITLE
add dqn pytorch

### PR DESCRIPTION
**Pull request by Asugawara**
In this PR, I added implements DQN in PyTorch benefitting from the TF.

[The experiment](https://github.com/deepmind/open_spiel/blob/36107100c820addf71b4f0fa50dca9462a385718/open_spiel/colabs/research_dqn_pytorch.ipynb) using breakthrough does not show a difference between PyTorch and TF.

Let me know I should experiment with other game if you need.

### CODE REVIEW

**Comment by lanctot** [comment]
> In this PR, I added implements DQN in PyTorch benefitting from the TF.

This is really cool! Thanks @Asugawara !

> The experiment using breakthrough does not show a difference between PyTorch and TF.
> 
> Let me know I should experiment with other game if you need.

Yeah, can you try it on Tic Tac Toe and inspect the output just to verify that it's doing something sensible? Also: does it solve catch?

**Comment by lanctot** [comment]
Note: I normally do the updates on Monday, but I will do one now so you have pytorch as a requirement and can enable the test.

**Comment by lanctot** [comment]
So......... how do you feel about NFSP next...? :)

**Comment by lanctot** [comment]
> Note: I normally do the updates on Monday, but I will do one now so you have pytorch as a requirement and can enable the test.

Done. Now you can rebase and have the pytorch requirement in the repos.

**Comment by michalsustr** [comment]
This is awesome! :)

Do you think you could create a super simple test, such as an MDP with three states: initial state is in the middle, going left gives reward of -1, going right gives reward of 1, and the neural network has to learn to go right? I think the algorithm should consistenly learn to get an expected reward higher than maybe 0.9 in just a couple of hundred iterations.

**Comment by michalsustr** [comment]
This is awesome! :)

Do you think you could create a super simple test, such as an MDP with three states: initial state is in the middle, going left gives reward of -1, going right gives reward of 1, and the neural network has to learn to go right? I think the algorithm should consistently learn to get an expected reward higher than maybe 0.9 in just a couple of hundred iterations.

**Comment by lanctot** [comment]
Great idea @michalsustr ! I will do this during the import. I will add a EFG game for this purpose and also add the same test to our TF dqn.

**Comment by michalsustr** [comment]
>  Do we know if there is an an standard analogous .mdp format?

I thought about that as well. I don't know about one, but maybe a subset of the dot language could be useful for this? Then you could easily visuallize it as well (just like the gambit trees).

**Comment by Asugawara** [comment]
Dr. @lanctot 
Thanks for your comments.

>Yeah, can you try it on Tic Tac Toe and inspect the output just to verify that it's doing something sensible? Also: does it solve catch?
I see! I’ll try to verify the game!

>Done. Now you can rebase and have the pytorch requirement in the repos.
Sorry for rushing you! I was going to add CMaketestlist after the update.
And then I'll rebase it!

>So......... how do you feel about NFSP next...? 🙂
Thank you for showing your roadmap! I'll implement NFSP next.


Dr. @michalsustr 
Thanks for your response!
Firstly, Spielviz is a very cool tool! 
It takes a lot of time for me to install.  difficult :) 

>Do you think you could create a super simple test, such as an MDP with three states
You’re smart! I didn’t come up with it!

>I will do this during the import. I will add an EFG game for this purpose and also add the same test to our TF dqn.
Should I wait for the test?

**Comment by lanctot** [comment]
> I was going to add CMaketestlist after the update. And then I'll rebase it!

Perfect, I did the update yesterday. I would like to see the test you added passing on Travis before I import this. 

> Should I wait for the test?

No need to wait for the MDP test Michal suggest, I will add it aftet I import this.

**Comment by Asugawara** [comment]
sorry! I made a mistake!

**Comment by lanctot** [comment]
> sorry! I made a mistake!

No worries. Your changes are contained in mostly the two files right? Maybe copy them somewhere and then start over with a new PR based on the current master branch?

**Comment by Asugawara** [comment]
Sorry again...
I open new PR...

and I close this PR...

**Comment by michalsustr** [comment]
> Firstly, Spielviz is a very cool tool! It takes a lot of time for me to install. difficult :)

Feel free to make an issue on spielviz if you encounter some bug or difficulty. It is still in development and not very high in my priorities, but if there is something that you're missing I will be happy to merge a pull request.
